Merge ShedToolLineage and StockLineage to ToolLineage

mvdbeek opened the pull request:
This means we don't rely on the Tool Shed metadata anymore and instead
deduce the tool lineage via the tool_id, which contains the version.
Comparing versions is done by comparing LooseVersions.
This works for toolshed and non-toolshed tools, restoring proper
grouping in the tool panel and version switches, even for tool shed
tools whose lineage is broken.

Should fix https://github.com/galaxyproject/galaxy/issues/4097 and would make fixing https://github.com/galaxyproject/galaxy/issues/1532 unnecessary (IMHO).

mvdbeek committed:
Replace ShedToolLineage with enhanced StockLineage

This means we don't rely on the Tool Shed metadata anymore and instead
deduce the tool lineage via the tool_id, which contains the version.
Version comparison then happens via LooseVersions.
This works for toolshed and non-toolshed tools, restoring proper
grouping in the tool panel and version switches, even for tool shed
tools whose lineage is broken.

mvdbeek committed:
Use correct tool version for ToolPanelManager.test_add_twice test to pass

mvdbeek committed:
Fix grouping tests by setting up correct versions

mvdbeek committed:
Keep StockLineage versions sorted

mvdbeek committed:
Remove unused next/previous version methods

jmchilton commented:
I couldn't love the ambition of this pull request more - this is fantastic. We should probably stipulate that there potentially **degenerate** cases where this provides a different lineage than what the tool shed would have provided before - which only effectively matters if a different tool really appears at the top of the lineage. I don't know if any actual repository exhibits such a lineage - but the fix in that case is pretty easy I think - just push out a new version of that tool with a new corrected absolute greater version and upgrade to that.

We can potentially delete even more client code that sets up those lineages and reads them from the server now right? That doesn't have to be done as part of this PR - but we should create an issue stipulate that we can do that I think.

mvdbeek commented:
> We should probably stipulate that there potentially degenerate cases where this provides a different lineage than what the tool shed would have provided before

Right, that could potentially happen. We could include this, and the solution to that problem in the release notes, right? And/Or we could go through the toolshed and see if that would happen -- I think the API exposes everything we need to do this remotely.

A corner-case may be versions like 1.0beta-1 appearing after 1.0 in the lineage, but we could solve this by switching to a version parser that implements [pep440](https://www.python.org/dev/peps/pep-0440/), like https://packaging.pypa.io/en/latest/version/#packaging.version.parse instead of LooseVersion.

> We can potentially delete even more client code that sets up those lineages and reads them from the server now right? That doesn't have to be done as part of this PR - but we should create an issue stipulate that we can do that I think.

We can definitely remove the lineage mako in the admin settings and remove the ToolVersionManager on the galaxy side. For now I think we may need to keep the ToolVersion and ToolVersionAssociation on the Tool Shed side for older Galaxy versions.

jmchilton commented:
@davebx do you know if there are any tool shed tests for tool migrations or put another way if this were to interfere with tool migrations do you think the tests would catch it?

nsoranzo commented:
You should also remove `tool_version_cache` from `test/unit/tools_support.py` .

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -436,7 +436,7 @@ def sa_session( self ):
     @property
     def tool_version( self ):
         """Return a ToolVersion if one exists for our id"""
```
I think now it returns a `StockLineage` object, so s/ToolVersion/StockLineage/ , right?

mvdbeek committed:
ToolVersion to StockLineage, drop ToolVersionCache

remnant from unittests (thx @nsoranzo!).

nsoranzo commented on lib/galaxy/model/tool_shed_install/__init__.py:
```diff
@@ -589,51 +589,11 @@ def __init__( self, id=None, create_time=None, tool_id=None, tool_shed_repositor
+        version_ids = lineage.get_versions()
         if reverse:
-            version_ids.reverse()
+            version_ids = reversed(version_ids)
```
I think that's not correct, [reversed](https://docs.python.org/2/library/functions.html#reversed) returns an iterator, while I think the expected return type of this method is a list.

nsoranzo commented on lib/galaxy/util/tool_version.py:
```diff
@@ -0,0 +1,19 @@
+def remove_version_from_guid(guid):
+
```
I'd remove this empty line.

nsoranzo commented on lib/galaxy/tools/toolbox/lineages/factory.py:
```diff
@@ -20,28 +11,39 @@ def __init__(self, app):
-        if from_toolshed:
-            lineage = ToolShedLineage.from_tool(self.app, tool)
-        else:
+        versionless_tool_id = remove_version_from_guid( tool_id)
```
I think you meant to also remove the space after the `(`

nsoranzo commented on lib/galaxy/tools/toolbox/lineages/factory.py:
```diff
@@ -20,28 +11,39 @@ def __init__(self, app):
             self.lineage_map[versionless_tool_id] = lineage
         if tool_id not in self.lineage_map:
             self.lineage_map[tool_id] = lineage
+        lineage.register_version(tool.version)
```
I don't think that's needed: `register_version(tool.version)` is already called inside `StockLineage.from_tool( tool )`, which is the only method used to create the `StockLineage` objects added to the `self.lineage_map` dictionary.

mvdbeek replied:
At first glance I agree with you, but removing this line breaks the lineages ... I'll have to dig into why that happens

nsoranzo replied:
Maybe a threading issue? See my comment below.

mvdbeek replied:
OK, I think this is what happens:
When a tool is loaded the first time, it will enter `self.lineage_map`, and have its version registered.
When a related tool is added, it will be fetched from `self.lineage_map` via the `versionless_tool_id`,  and not run through `ToolLineage.from_tool`. So we need to add the current tool's version via `lineage.register_version(tool.version)`. I'll add a comment to clarify this.

nsoranzo replied:
I see! Then you can move this up a bit changing lines 18-19 above to:
```python
        if lineage:
            lineage.register_version(tool.version)
        else:
            lineage = ToolLineage.from_tool(tool)
```

nsoranzo commented on lib/galaxy/util/tool_version.py:
```diff
@@ -0,0 +1,19 @@
+    return guid[:last_slash]
+
+
+def get_version_from_guid(guid):
```
This method is unused.

nsoranzo commented on lib/galaxy/model/tool_shed_install/__init__.py:
```diff
@@ -589,51 +589,11 @@ def __init__( self, id=None, create_time=None, tool_id=None, tool_shed_repositor
-        __descendants( app, self )
-        return tool_versions
-
     def get_version_ids( self, app, reverse=False ):
```
This method is not used any more.

mvdbeek replied:
It is still used by https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/controllers/admin_toolshed.py#L2066, but I think we can replace this.

nsoranzo replied:
Sorry I missed that! Up to you if you want to replace it or leave it as it is.

nsoranzo commented on lib/galaxy/webapps/galaxy/controllers/admin.py:
```diff
@@ -455,13 +455,20 @@ class ToolVersionsColumn( grids.TextColumn ):
+            if tool:
+                tool_id = tool.id
+                lineage = trans.app.toolbox._lineage_map.get(tool_id)
+                if lineage:
```
If `lineage` is `None`, the method returns an empty string. I don't think that should ever happen, so it may be worth to remove this `if` and let an `Exception` be raised on the next line if something is really broken.

nsoranzo reviewed (commented):
Amazing refactoring @mvdbeek! Hopefully this will solve the duplications of TS tools in the toolbox.

mvdbeek committed:
Remove duplication between ToolLineage and TVM

Getting tool versions and tool ids is now done through the tool lineage.
Also assume that the lineage can always be found for a given loaded tool in
the ToolVersionListGrid class. Thanks @nsoranzo for these suggestions!

mvdbeek committed:
Remove unused `get_version_from_guid` function

nsoranzo commented:
Should `lib/galaxy/tools/toolbox/lineages/stock.py` be merged in `lib/galaxy/tools/toolbox/lineages/interface.py` leaving only the `ToolLineage` class?

nsoranzo commented on lib/galaxy/webapps/galaxy/controllers/admin.py:
```diff
@@ -455,13 +455,18 @@ class ToolVersionsColumn( grids.TextColumn ):
-                    tool_ids_str += '%s<br/>' % tool_id
+            tool = toolbox._tools_by_id.get(tool_version.tool_id)
+            if tool:
+                tool_id = tool.id
```
This line can be removed.

mvdbeek committed:
Remove unused tool_id variable

mvdbeek committed:
Rename StockLineage to ToolLineage and merge stock.py contents into interface.py

nsoranzo commented on lib/galaxy/tools/toolbox/lineages/interface.py:
```diff
@@ -50,3 +37,59 @@ def to_dict(self):
+            if tool_id not in lineages_by_id:
+                lineages_by_id[ tool_id ] = ToolLineage( tool_id )
+        lineage = lineages_by_id[ tool_id ]
+        lineage.register_version( tool.version )
```
I'm not sure that modifying `lineage` here by calling its `register_version()` method is thread-safe. What about rewriting lines 67-71 as:
```python
        with ToolLineage.lock:
            lineage = lineages_by_id.get(tool_id, ToolLineage(tool_id))
            lineage.register_version( tool.version )
            lineages_by_id[ tool_id ] = lineage
```
?

mvdbeek replied:
Hmm, that didn't seem to be enough ...

nsoranzo commented:
I think we can also remove line 418-419 in `lib/galaxy/tools/__init__.py`

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -441,8 +441,8 @@ def sa_session( self ):
 
     @property
     def tool_version( self ):
```
Crazy idea: would it make sense to rename this property to `lineage` and remove:
```python
tool.lineage = tool_lineage
```
from `AbstractToolBox._load_tool_tag_set()` method in `lib/galaxy/tools/toolbox/base.py` ?

I think this property is only used and would need to be renamed:
- in the next method `tool_versions`
- in this file at line1835
- in `lib/galaxy/webapps/galaxy/controllers/admin_toolshed.py` at line 2064.

mvdbeek replied:
absolutely, I think that makes it clearer.

mvdbeek committed:
Remove unused lineage_ids from Tool class

nsoranzo reviewed (commented):
Thanks @mvdbeek for the changes!

mvdbeek committed:
Clarify when and why tool version is being registered

nsoranzo reviewed (approved):
Not sure if you also want to implement my "crazy idea" before merge, but :+1: from me anyway.

mvdbeek commented:
I have been playing with this for a while now, the problem is that if we do the renaming operation we need to have a working `app.toolbox._lineage_map` object in the unit tests ... and that seems to be tricky.

mvdbeek committed:
Move tool.tool_version to tool.lineage

This change makes it obvious that a Lineage object will be returned.

Also removes two unused methods from ToolLineageVersion.

mvdbeek commented:
@nsoranzo I have tried to go with the spirit of your 'crazy idea'. If 41d739b is fine for you this is ready to be merged.

nsoranzo commented on lib/galaxy/webapps/galaxy/controllers/admin_toolshed.py:
```diff
@@ -2059,11 +2059,12 @@ def view_tool_metadata( self, trans, repository_id, tool_id, **kwd ):
+                        tool = trans.app.toolbox.load_tool( os.path.abspath( tool_config ), guid=tool_metadata[ 'guid' ] )
+                        if tool:
+                            tool._lineage = trans.app.toolbox._lineage_map.register( tool )
+                    tool_lineage = tool.lineage.get_version_ids(reverse=True)
```
This line should be indented as the line above, i.e. executed only if `tool` exists.

mvdbeek replied:
Then we wouldn't get the lineage if the tool was already in the toolbox.
I have added another `if tool:` statement in 1e11c46c348ad65835bda2cb5468c7b3e3b2e303

mvdbeek committed:
Access tool.lineage only if tool was loaded

nsoranzo commented on test/api/test_dataset_collections.py:
```diff
@@ -1,4 +1,5 @@
 import json
+import time
 import tarfile
```
Travis complains about the import order here (assuming you added the changes to this file on purpose).

mvdbeek replied:
nope ... :(, that came from #4222

nsoranzo reviewed (approved):
@jmchilton Ready for merge?

jmchilton commented:
Great - thanks a bunch for the excellent implementation @mvdbeek and the detailed review @nsoranzo!

guerler commented:
I like this PR a lot but I noticed an issue when loading a previously imported workflow and I think it might be related.

```
`File '/Users/guerler/galaxy/lib/galaxy/web/framework/decorators.py', line 54 in call_and_format
  rval = func( self, trans, *args, **kwargs )
File '/Users/guerler/galaxy/lib/galaxy/webapps/galaxy/controllers/workflow.py', line 639 in load_workflow
  return workflow_contents_manager.workflow_to_dict( trans, stored, style="editor" )
File '/Users/guerler/galaxy/lib/galaxy/managers/workflows.py', line 324 in workflow_to_dict
  return self._workflow_to_dict_editor( trans, stored )
File '/Users/guerler/galaxy/lib/galaxy/managers/workflows.py', line 448 in _workflow_to_dict_editor
  config_form = module.get_config_form()
File '/Users/guerler/galaxy/lib/galaxy/workflow/modules.py', line 708 in get_config_form
  return self.tool.to_json( self.trans, incoming, workflow_building_mode=True )
File '/Users/guerler/galaxy/lib/galaxy/tools/__init__.py', line 1834 in to_json
  tool_versions = self.lineage.tool_versions
AttributeError: 'NoneType' object has no attribute 'tool_versions'`
```
I can debug this more and provide additional information if needed.

mvdbeek commented:
Hmm, that is interesting. Once a tool is loaded into the tool panel it should get a lineage. Any chance this workflow is missing a tool ?

guerler commented:
Yes it is missing tools.

mvdbeek commented:
OK, I think we can just be a bit more defensive in determining the `tool_versions`. I'll try to reproduce this.

mvdbeek commented:
@guerler can you try https://github.com/mvdbeek/galaxy/commit/ad5431cf551cc3b3b18425405a151ed91e169971 and see if that is all that is needed? Also, any idea how I can reproduce this? If I just uninstall a tool the workflow export fails with
```
File '/Users/mvandenb/src/galaxy/lib/galaxy/managers/workflows.py', line 585 in _workflow_to_dict_export
  if module.tool.tool_shed_repository:
AttributeError: 'NoneType' object has no attribute 'tool_shed_repository'
```

but that is not a new error ...

guerler commented:
Yes, this fixes the issue I mentioned above.
